### ISSUE TITLE
False positive for undefined variable with assignment expression

### ISSUE THREAD

**Issue by hauntsaninja** [PyCQA/pylint#3688]
### Steps to reproduce
```
~/tmp λ cat test4.py
def i(am: str = (the := "walrus")) -> str:
    return "goo goo g'joob"
print(the)
```
### Current behavior
```
~/tmp λ pylint -E test4.py
************* Module test4
test4.py:5:6: E0602: Undefined variable 'the' (undefined-variable)
```
### Expected behavior
pylint recognises that `the` is defined. Alternatively, pylint yells at me for doing this in the first place :-)

### pylint --version output
```
~/tmp λ pylint --version
pylint 2.5.3
astroid 2.4.2
Python 3.8.3 (default, May 27 2020, 20:54:22) 
[Clang 11.0.3 (clang-1103.0.32.59)]
```

Pretty minor, especially given that assignment expressions aren't a priority for pylint (https://github.com/PyCQA/pylint/issues/3275#issuecomment-559122266).

**Comment by bersbersbers** [comment]
Here's another case. Interestingly,
```
x = next(i for i in range(10) if (y := i * i))
print(y)
```
raises this false-positive, while
```
next(i for i in range(10) if (y := i * i))
print(y)
`` 
does not.

**Comment by egormm** [comment]
A couple more interesting cases with false-positive behaviour:
```
print({'key': value if (value := 'something') else 'anything'})
```
 raises `used-before-assigment`, while 
```
value = val if (val := 'something') else 'anything'
print({'key': value})
```
doesn't. Excepting behaviour - no errors.
Same with classes and dataclasses:
```
class Dummy:
    def __init__(self, value):
        self.value = value

dummy = Dummy(value=val if (val := 'something') else 'anything')
print(dummy.value)
```
raises `used-before-assigment`, while 
```
class Dummy:
    def __init__(self, value):
        self.value = value

value = val if (val := 'something') else 'anything'
dummy = Dummy(value)
print(dummy.value)
```
doesn't. Excepting behaviour - no errors.
